### ISSUE TITLE
Move translation out of Transifex

### ISSUE THREAD

**Issue by frimelle** [inventaire/inventaire#67]
Might be considered to use https://translatewiki.net/ for the translation of inventaire to have it there with a lot of other cool projects for people to join the translation efforts.

**Comment by maxlath** [comment]
@frimelle interesting! I love the fact that anyone can contribute without creating an account first; the typical workflow on Transifex being:
- someone wants to help translating and find the Transifex project, create an account and make a request
- I see the notification one week later and validate the request to join a given language team
- the person long forgot about it and, in the case she got a notification for the acceptance, is probably not in the flow any more

A few questions:
- do you know how the API workflow works? that's probably the biggest obstacle: re-integrating strings imports and exports
- are the strings mutualized between the projects? like, if a string is not translated in inventaire in Dutch but the same string is available in Dutch in an other project, it will be imported? that would be beautiful :D

**Comment by almereyda** [comment]
We may offer [weblate.allmende.io](https://weblate.allmende.io) at some point https://github.com/TransforMap/weblate.transformap.co/issues/1

**Comment by maxlath** [comment]
we are well installed on Transifex at the moment, and despite it's non-opensourceness, the cost of moving to another platform is too high: closing the issue for now

**Comment by maxlath** [comment]
https://github.com/WeblateOrg/weblate

**Comment by almereyda** [comment]
If you please we could migrate weblate.transformap.co to weblate.allmende.io
and offer that one instead.

**Comment by almereyda** [comment]
@maxlath Newest Weblate 2.20 is now running on https://weblate.allmende.io

Just sayin'.

**Comment by maxlath** [comment]
re-opening as Transifex might be hurting contributions by making progresses hard to follow (maybe contributors stats are a premium?!?)
poke @jum-s

**Comment by maxlath** [comment]
requirements for the new translation platform:
* libre
* clear interface
* translations can be extracted from an API, preferably without authentification
* statistics on contributions and evolution of translations, so that we can see when there is a big effort going on, offer help, and thank the contributor

anything else?

**Comment by maxlath** [comment]
some [input from @wetneb](https://mamot.fr/@pintoch/102121968201990744):
> In my experience Weblate is just fantastic, it feels very professional. And Translatewiki brings you a lot of very keen translators who will translate your app in languages you had no idea existed!

**Comment by nemobis** [comment]
This is how to add a project on translatewiki.net:
https://translatewiki.net/wiki/Translating:New_project

We can work on it now that we're in Stockholm, happy to help.

You can also see how it's going with Dissemin, mentioned by wetneb above:
https://translatewiki.net/wiki/Translating:Dissemin

**Comment by vk0xOrg** [comment]
> hum statistics, unsure its a key feature to me since it individualizes contributions (rewarding/badges).
> framafolks use [Zanata](https://trad.framasoft.org/project/view/contributopia) under copyright by Red Hat
> actually [Weblate](https://weblate.org) suggested by @almereyda earlier looks like this
> ![selection_007](https://user-images.githubusercontent.com/5363918/50973416-1f811700-14e9-11e9-8315-bf5b270a0475.png)
> can that fit into "clear interface" specs
> there is a django alternative: [Pootle](https://pootle.translatehouse.org/) ![screenshot duck search](https://camo.githubusercontent.com/6184073e1567a9abf3844cb9419ddb33e75f008c/68747470733a2f2f6475636b6475636b676f2e636f6d2f3f713d706f6f746c652b7472616e736c6174696f6e26743d66666162266961723d696d61676573266961783d696d616765732669613d696d61676573266961693d6874747073253341253246253246612e6673646e2e636f6d253246636f6e25324661707025324670726f6a2532467472616e736c61746525324673637265656e73686f7473253246506f6f746c65253235323045646974696e672e706e67) which seem to me a little less "clear" than weblate

---

@jum-s since then framafols have been moving to Weblate [annoncement](https://trad.framasoft.org/project/view/framasoft?dswid=7460)
https://weblate.framasoft.org/projects/framasoft

![Screenshot at 2020-07-26 13-35-21](https://user-images.githubusercontent.com/45686717/88476988-1168df00-cf45-11ea-90df-1f72df25f52c.png)

**Comment by vk0xOrg** [comment]
@maxlath I've sent an joining request on Transifex to join/create team for Ukrainian translation. I don't know how it works and maybe you have not been notified for the new requests (taking into account your comment re receiving notifications one week later after it was requested).
So maybe this is one more point to move towards in this question and to switch translation platform.

**Comment by maxlath** [comment]
I would really like to be able to `PUT /${resource}/${language}/${key} {"value":"foo"}` (or something approaching). I couldn't find how to do that with Transifex, anyone knowing if that's possible with Weblate?

**Comment by wetneb** [comment]
It seems possible:
https://docs.weblate.org/en/latest/api.html#post--api-translations-(string-project)-(string-component)-(string-language)-units-

**Comment by jum-s** [comment]
About authentication, inventaire is now an [oauth provider](https://github.com/inventaire/inventaire/pull/487). However, weblate seems to have a [defined set of connectors](https://docs.weblate.org/en/weblate-4.6.2/admin/auth.html#social-authentication) based on [Python Social Auth](https://github.com/WeblateOrg/weblate/issues/5629) (although it does seem to [allow custom provider](https://python-social-auth.readthedocs.io/en/latest/use_cases.html#signup-by-oauth-access-token), i havent found any custom provider implementation on weblate docs (only GAFAM + gitlab solutions)).
Maybe its possible to setup a gitlab which oauth from inventaire, and then a weblate which oauth from gitlab ?

**Comment by almereyda** [comment]
I think here we need to differentiate between OAuth2 and OpenID Connect (OIDC), which target different use cases:

- The prior is for authorization (authZ) (of API requests) and
- the latter is for authentication (authN) (of user identities) through a combination of OpenID and OAuth2.

As https://github.com/inventaire/inventaire/issues/84#issuecomment-898881110 shows the implementation of #487 apparently already works as a login method, one could try leveraging any of

* https://github.com/python-social-auth/social-core/blob/master/social_core/backends/oauth.py
* https://github.com/python-social-auth/social-core/blob/master/social_core/backends/open_id_connect.py

for bridging to the Inventaire "identity provider" (IDP) by configuring them according to spec.

Any federated login to GitLab through Inventaire sounds like an unneccessary round trip to me.

---

<details>

<summary>Having said this, please note:</summary>

To my defense, I'm regularily logging in to our very own https://lab.allmende.io through the federated GitHub login, because they don't invalidate my session as quickly, so I don't need to type a password, and LibreHosters also use federated login for https://lab.libreho.st through login.libreho.st (Keycloak) via git.coop, IndieHosters or weho.st

While your workaround seems possible, I am not perfectly sure if it is worthwhile.

</details>

**Comment by jum-s** [comment]
some conversation about this topic can be found at https://mastodon.ar.al/@aral/109222855802977070

**Comment by jum-s** [comment]
Since weblate is quite popular and framasoft offered us a little space on their instance (https://weblate.framasoft.org/projects/inventaire/ Very nice of them! 🎉 ) we are opting for not hosting this service which would nicely ease maintenance, while moving the stack toward more free software.

2 questions to be discussed before starting:

### Merging full & short keys into one component

Today, several [components](https://docs.weblate.org/en/latest/admin/projects.html#component-configuration) are defined ([here](https://github.com/inventaire/inventaire-i18n/tree/master/original)). While it could make sense to keep server keys in its own component (since translations will not be consumed by the client repo), merging short and full keys would ease translators experience. This would require to extend shortkey formatting (interpolation) to fullkey, but whom can do more, can do less i guess.

wikidata.properties_list seems to have a nice isolation in its own component too. 


### Version control vs API ?

When pulling new translation keys ([step 3](https://docs.weblate.org/en/latest/admin/continuous.html#update-vcs)), Weblate can either rely on version control (VCS) or using a good ol REST API. While I can see an [advantange](https://docs.weblate.org/en/latest/admin/continuous.html#update-vcs) to have fully automated hook updates (based on the event to push a new commit (see [GH webhook](https://docs.weblate.org/en/latest/admin/continuous.html#github-setup))), i can see some pitfalls (i would be opposed to [couple weblate to our current code host](https://docs.weblate.org/en/latest/admin/continuous.html#gitlab-setup) (since it might change in the future)) to rely on CI (for example how to warn if something went wrong). I would rather rely on some homemade scripts (see [Weblate’s REST API](https://docs.weblate.org/en/latest/api.html)) and use the api.

If anyone would like to share their awesome VCS experience i would be happy to read it.

### Steps to migrate
Here are some step i could forsee (feel free to edit this post and add missed points):

**Prerequisites**
- [ ] ensure all languages are on Weblate
- [ ] answer/decide the 2 questions above

**Migration**
- [ ] pull transifex translations
- [ ] create and configure adequate components
- [ ] update `inventaire-i18n/scripts` with weblate flow
- [ ] run scripts and transfer currents translations

**Cleanup**
- [ ] delete Transifex project at https://www.transifex.com/inventaire/inventaire/settings/
- [ ] cleanup `transifexify` package in `inventaire-i18n`

**Documentation**
- [ ] explicitly mention Weblate in `inventaire-i18n/README.md`.
- [ ] update subdomain redirection : `translate.inventaire.io`
- [ ] update wiki [transifex occurences](https://wiki.inventaire.io/w/index.php?search=transifex&title=Special%3ASearch&go=Go)

**Communication**
- [ ] warn current transifex people, by adding a "new discussion" in transifex team (https://www.transifex.com/inventaire/teams/37854/discussions/) 
- post a message on [communication channels](https://wiki.inventaire.io/wiki/Communication_channels)

**Comment by almereyda** [comment]
For TransforMap we used https://weblate.transformap.co/ which seems quite resilient up until today. This gave us a nice log of contributions https://github.com/TransforMap/transformap-viewer-translations/commits/master and simple JSON files to ingest (from GitHub CDN back then).

The Weblate VCS integrations are supported and proven stable since many years. You will be happy to hear that there is first class support not only for GitHub, but also GitLab and Gitea, which should cover the most suitable of the available alternatives. https://docs.weblate.org/en/latest/vcs.html If you are indeed looking into more radical alternatives, like https://sr.ht/, then indeed implementing and maintaining a custom scripted API integration pipeline as a loosely coupled CI workflow in inventaire-i18n (at GitHub ^_^) may seem a good fit.

Else there remains the possibility to use the existing upstream offers, and with that reduce maintained lines of code.

**Comment by maxlath** [comment]
After some troubles getting into Weblate at first (many options and I was probably too much looking for a 1-to-1 mapping with Transifex), I'm happy to discover that it can be used as a fancy GUI on top of git (:star_struck:), making fetching data as simple as
```sh
git add remote weblate https://weblate.framasoft.org/git/inventaire/server/
git pull weblate master
```
:tada: 

[Weblate could push directly to Github](https://docs.weblate.org/en/weblate-4.16.4/admin/continuous.html#push-changes), as https://hosted.weblate.org does with the help of a dedicated user, but it doesn't look like this is setup for https://weblate.framasoft.org/ (or I couldn't find the user). In the meantime, fetching translation commits from the Weblate remote seems like a workable option, so I'll be looking into that if there is no better suggestion.

**Comment by nemobis** [comment]
Il 30/03/23 01:32, maxlath ha scritto:
> In the meantime, fetching translation commits from the Weblate remote seems like a workable option,

Note that to reduce the need for manual resolution of conflicts Weblate 
recommends to do this automatically and quite frequently.

Speaking of which, translatewiki.net does that for you. :) </shameless plug>

**Comment by wetneb** [comment]
If @nemobis is shamelessly advertising translatewiki.net then I have to intervene and tell you once again my limitless love for Weblate and its amazing features, including the Git(hub) integration which makes it a breeze. It has the added advantage that your translators will be correctly marked as (co)-authors of the corresponding commits, which is great to recognize those contributions in the activity of your repository (instead of having a faceless bot take the credit for all this work).

Also I wonder why hosted.weblate.org is not an option for you? (Encouraging decentralization by going for a different provider is definitely laudable. But oh my, the GitHub integration!!!)

**Comment by maxlath** [comment]
We did it, we are on Weblate, and everything is going smoothly :tada: 
https://weblate.framasoft.org/engage/inventaire/

I couldn't setup the push from weblate.framasoft.org to Github, as there is no official Github user for that Weblate instance, but some user is already using the associated SSH key. Instead, we have hooks to [pull Weblate commits before commiting](https://github.com/inventaire/inventaire-i18n/commit/4e2e16c) and before building i18n assets. That might just be fine for our use-case as i18n strings are in a dedicated repository (https://github.com/inventaire/inventaire-i18n).

As for why sticking to weblate.framasoft.org rather than hosted.weblate.org: we are in pretty good company there :) https://weblate.framasoft.org/projects/

I let this issue open the time to check the last boxes in @jum-s comment above.
